File: LeetCode-Solutions/0684-redundant-connection/solution.cpp
class Solution {
public:
    int root ( int u, vector<int>& cha ) {
        if ( cha[u] != u ) cha[u] = root(cha[u],cha);
        return cha[u];
    }
    void join( int  x , int  y, vector<int>& cha) {
        int  u = root(x,cha);
        int  v = root(y,cha);
        cha[v] = u;
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();
        vector<int> cha(n + 1);
        for ( int i = 0 ; i <= n ; ++i ) {
            cha[i] = i;
        }
        for ( auto e : edges ) {
            int u = root(e[0],cha);
            int v = root(e[1],cha);
            if ( u == v ) {
                return e;
            } else {
                join(e[0],e[1],cha);
            }
        }
        return {};
    }
};

File: LeetCode-Solutions/0684-redundant-connection/solution.java
class Solution {
    private int root(int u, int[] parent) {
        if (parent[u] != u) {
            parent[u] = root(parent[u], parent);
        }
        return parent[u];
    }

    private void join(int x, int y, int[] parent) {
        int u = root(x, parent);
        int v = root(y, parent);
        parent[v] = u;
    }

    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        int[] parent = new int[n + 1];
        
        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }
        
        for (int[] edge : edges) {
            int u = root(edge[0], parent);
            int v = root(edge[1], parent);
            
            if (u == v) {
                return edge;
            } else {
                join(edge[0], edge[1], parent);
            }
        }
        return new int[0];
    }
}

